packages/web3-eth-abi/src/types.ts

```diff
@@ -1,57 +1,187 @@
 }
 
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiParameter = {
+	readonly name: string;
+	readonly type: string;
+	readonly baseType?: string;
+	readonly indexed?: boolean;
+	readonly components?: ReadonlyArray<AbiParameter>;
+	readonly arrayLength?: number;
+	readonly arrayChildren?: ReadonlyArray<AbiParameter>;
+};
+
 type FragmentTypes = 'constructor' | 'event' | 'function';
 
-export interface JsonAbiBaseFragment {
-	name?: string;
-	type: FragmentTypes;
-	inputs?: Array<JsonAbiParameter>;
-}
+export type AbiBaseFragment = {
+	readonly type: FragmentTypes;
+};
 
-export interface JsonAbiConstructorFragment extends JsonAbiBaseFragment {
-	type: 'constructor';
-	stateMutability: 'nonpayable' | 'payable';
-}
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiConstructorFragment = AbiBaseFragment & {
+	readonly type: 'constructor';
+	readonly stateMutability: 'nonpayable' | 'payable';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+};
 
-export interface JsonAbiFunctionFragment extends JsonAbiBaseFragment {
-	type: 'function';
-	stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
-	outputs?: Array<JsonAbiParameter>;
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiFunctionFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'function';
+	readonly stateMutability: 'nonpayable' | 'payable' | 'pure' | 'view';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly outputs: ReadonlyArray<AbiParameter>;
 
 	// legacy properties
-	constant?: boolean;	// stateMutability == 'pure' or stateMutability == 'view'
-	payable?: boolean; // stateMutability == 'payable'
-}
+	readonly constant?: boolean; // stateMutability == 'pure' or stateMutability == 'view'
+	readonly payable?: boolean; // stateMutability == 'payable'
+};
 
-export interface JsonAbiEventFragment extends JsonAbiBaseFragment {
-	type: 'event';
-	anonymous?: boolean;
-}
+// https://docs.soliditylang.org/en/latest/abi-spec.html#json
+export type AbiEventFragment = AbiBaseFragment & {
+	readonly name: string;
+	readonly type: 'event';
+	readonly inputs: ReadonlyArray<AbiParameter>;
+	readonly anonymous?: boolean;
+};
 
 // https://docs.soliditylang.org/en/latest/abi-spec.html#json
-export type JsonAbiFragment =
-	| JsonAbiConstructorFragment
-	| JsonAbiFunctionFragment
-	| JsonAbiEventFragment;
+export type AbiFragment = AbiConstructorFragment | AbiFunctionFragment | AbiEventFragment;
+
+export type ContractAbi = ReadonlyArray<AbiFragment>;
+
+export type AbiInput = string | AbiParameter | { readonly [key: string]: unknown };
+
+export type FilterAbis<Abis extends ContractAbi, Filter, Abi = Abis[number]> = Abi extends Filter
+	? Abi
+	: never;
+
+type _TypedArray<Type, Size extends string> = Size extends ''
+	? Type[]
+	: FixedSizeArray<Type, ConvertToNumber<Size>>;
+
+export type PrimitiveAddressType<Type extends string> = Type extends `address[${infer Size}]`
```

Comment:
As discussed , could you add doc comments or in readme file list of supported types checking.

Reply:
All the solidity types which can be used in `ABI` are supported as per my knowledge. So seems no need to mention which types are supported.